core/server/src/main/java/alluxio/master/file/FileSystemMaster.java

```diff
@@ -1520,46 +1551,106 @@ public long loadMetadata(AlluxioURI path, boolean recursive)
+      AccessControlException, FileAlreadyCompletedException, InvalidFileSizeException {
+    if (mInodeTree.inodePathExists(path)) {
+      return mInodeTree.getInodeByPath(path).getId();
+    }
+    AlluxioURI ufsUri = resolution.getUri();
+    UnderFileSystem ufs = resolution.getUfs();
+
+    long ufsBlockSizeByte = ufs.getBlockSizeByte(ufsUri.toString());
+    long ufsLength = ufs.getFileSize(ufsUri.toString());
+    // Metadata loaded from UFS has no TTL set.
+    CreateFileOptions createFileOptions =
+        CreateFileOptions.defaults().setBlockSizeBytes(ufsBlockSizeByte)
+            .setRecursive(options.isCreateAncestors()).setMetadataLoad(true).setPersisted(true);
+    try {
+      long fileId = createFile(path, createFileOptions);
+      CompleteFileOptions completeOptions = CompleteFileOptions.defaults().setUfsLength(ufsLength);
+      completeFile(path, completeOptions);
+      return mInodeTree.getInodeById(fileId).getId();
+    } catch (FileAlreadyExistsException e) {
+      LOG.error("FileAlreadyExistsException seen unexpectedly.", e);
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
    * Loads metadata for the directory identified by the given path from UFS into Alluxio. This does
    * not actually require looking at the UFS path.
+   * It is a no-op if the directory exists and is persisted.
    *
    * @param path the path for which metadata should be loaded
-   * @param recursive whether parent directories should be created if they do not already exist
+   * @param options the load metadata options
    * @return the file id of the loaded directory
-   * @throws FileAlreadyExistsException if the object to be loaded already exists
    * @throws InvalidPathException if invalid path is encountered
    * @throws IOException if an I/O error occurs   *
    * @throws AccessControlException if permission checking fails
    * @throws FileDoesNotExistException if the path does not exist
    */
   @GuardedBy("mInodeTree")
-  private long loadDirectoryMetadata(AlluxioURI path, boolean recursive)
-      throws IOException, FileAlreadyExistsException, InvalidPathException, AccessControlException,
-      FileDoesNotExistException {
-    CreateDirectoryOptions options =
+  private long loadDirectoryMetadata(AlluxioURI path, LoadMetadataOptions options)
+      throws IOException, InvalidPathException, AccessControlException, FileDoesNotExistException {
+    if (mInodeTree.inodePathExists(path)) {
+      Inode inode = mInodeTree.getInodeByPath(path);
+      if (inode.isPersisted()) {
+        return inode.getId();
+      }
+    }
+    CreateDirectoryOptions createDirectoryOptions =
         CreateDirectoryOptions.defaults().setMountPoint(mMountTable.isMountPoint(path))
-            .setPersisted(true).setRecursive(recursive).setMetadataLoad(true);
-    InodeTree.CreatePathResult result = createDirectory(path, options);
+            .setPersisted(true).setRecursive(options.isCreateAncestors()).setMetadataLoad(true)
```

Comment:
optional: rename the CreatePathOptions field `mRecursive` to `mCreateAncestors` ... or at least add a TODO to do that